rules/apple_library.bzl

```diff
@@ -0,0 +1,673 @@
+            additional_objc_infos = swift_objc_infos,
+            alwayslink = ctx.attr.alwayslink,
+            deps = deps,
+            feature_configuration = feature_configuration,
+            module_context = module_context,
+            libraries_to_link = swift_objc_link_inputs,
+            user_link_flags = linkopts,
+        )
+
+    cc_toolchain = find_cpp_toolchain(ctx)
+    cc_feature_configuration = cc_common.configure_features(
+        ctx = ctx,
+        cc_toolchain = cc_toolchain,
+        requested_features = ctx.features,
+        unsupported_features = ctx.disabled_features,
+    )
+
+    # This is a different VFS than the otherone
+    # Note: this needs to go here, in order to virtualize the extended module
+    module_map = ctx.files.module_map[0] if ctx.files.module_map else None
+    if module_map and generated_header_file:
+        module_mapo = _extend_modulemap(ctx, generated_header_file)
+        module_map = module_mapo
+
+    swiftmodules = [module_context.swift.swiftmodule] if module_context else []
+    objc_vfs = make_vfsoverlay(
+        ctx,
+        hdrs = objc_hdrs + ([generated_header_file] if generated_header_file else []),
+        output = ctx.actions.declare_file(ctx.attr.name + "_objc.yaml"),
+        module_map = [module_map] if module_map else ctx.files.module_map,
+        framework_name = ctx.attr.module_name,
+        # We might need to pass in .swiftinterface files here as well
+        # esp. if the error is `swift declaration not found` for some module
+        swiftmodules = swiftmodules,
+        private_hdrs = ctx.files.private_hdrs,
+        has_swift = len(ctx.files.swift_srcs) > 0,
+        merge_vfsoverlays = fw_dep_vfsoverlays,
+        extra_search_paths = None,
+    )
+
+    vfs_objc_copts = [
+        "-ivfsoverlay{}".format(objc_vfs.vfsoverlay_file.path),
+        "-F{}".format(VFS_OVERLAY_FRAMEWORK_SEARCH_PATH),
+    ]
+
+    # Note: here we create a generated headermap for the objc library.
+    swift_hmap = ctx.actions.declare_file(ctx.attr.name + "gen.hmap")
+    hmap.make_hmap(
+        actions = ctx.actions,
+        headermap_builder = ctx.executable._headermap_builder,
+        output = swift_hmap,
+        namespace = hmap_namespace,
+        # This might be a bug - we'll feed the direct headers to the hmap
+        hdrs_lists = [module_context.clang.compilation_context.direct_headers] if module_context else [],
+    )
+    label = ctx.label
+
+    objc_srcs = []
+    for h in ctx.attr.objc_srcs:
+        if DefaultInfo in h:
+            objc_srcs.extend(h[DefaultInfo].files.to_list())
+        else:
+            objc_srcs.append(h)
+
+    objc_non_arc_srcs = []
+
+    # TODO: these are smashed as 1 target
+    for h in ctx.attr.non_arc_srcs:
+        if DefaultInfo in h:
+            objc_non_arc_srcs.extend(h[DefaultInfo].files.to_list())
+        else:
+            objc_non_arc_srcs.append(h)
+
+    objc_cc_info = None
```

Comment:
The ideas are here - but is should be split into a number of helper functions instead of 1 massive one 🔨